inference-engine/src/legacy_api/include/legacy/details/ie_cnn_network_iterator.hpp

```diff
@@ -32,25 +33,87 @@ CNNNetworkIterator {
     IE_SUPPRESS_DEPRECATED_START
 
     std::unordered_set<CNNLayer*> visited;
-    std::list<CNNLayerPtr> nextLayersTovisit;
+    std::list<CNNLayerPtr> nextLayersToVisit;
     InferenceEngine::CNNLayerPtr currentLayer;
-    ICNNNetwork* network = nullptr;
+    const ICNNNetwork* network = nullptr;
+    bool oldApproach = false;
 
-    void init(const ICNNNetwork* network) {
+    void init(const ICNNNetwork* net) {
+        network = net;
         if (network == nullptr) THROW_IE_EXCEPTION << "ICNNNetwork object is nullptr";
-        // IE_ASSERT(dynamic_cast<const details::CNNNetworkImpl*>(network) != nullptr);
+
+        OutputsDataMap outputs;
+        network->getOutputsInfo(outputs);
+
         InputsDataMap inputs;
         network->getInputsInfo(inputs);
-        if (!inputs.empty()) {
-            auto& nextLayers = getInputTo(inputs.begin()->second->getInputData());
-            if (!nextLayers.empty()) {
-                currentLayer = nextLayers.begin()->second;
-                nextLayersTovisit.push_back(currentLayer);
-                visited.insert(currentLayer.get());
+
+        auto get_consumers = [](const CNNLayerPtr& node) -> std::vector<CNNLayerPtr> {
+            std::vector<CNNLayerPtr> consumers;
+            for (const auto & output : node->outData) {
+                for (const auto &consumer : getInputTo(output)) {
+                    consumers.push_back(consumer.second);
+                }
             }
+            return consumers;
+        };
+        auto bfs = [&](const CNNLayerPtr& start_node, bool traverse_via_outputs = false) {
```

Comment:
if this bfs ported from v7 serializer ? as i remember it never worked in real networks, always missed some layers in iterations.

Reply:
@esmirno I took it from the `SortTopologically` function. As I know we actively use it in the IE plugins.

Reply:
Moreover, as I remember @GlebKazantaev fixed this method.